### DISCUSSION TITLE
Automatic Events Log analyser - for observations & troubleshooting

### DISCUSSION THREAD

**Discussion by nigelkeirross** [NightscoutFoundation/xDrip#2786] [Ideas]
Has there been any consideration to building into xDrip+ the facility to analyse the events log. As xDrip has more features, the number & different entries in the log can be somewhat intimidating, and can sometime lead to repeated questions, with only partial information provided. Not all users will understand the significance of the information in the log, and the sheer volume of data can overwhelm, unless one knows precisely what to look for.
Perhaps the first stage would be to have the tool provide some simple observations, such as:
 - You are getting lots of zero values in your raw data.
 - You have good BT connectivity with your Libre2 as you are getting data each minute.
 
And the next stage could be t provide some diagnostics, or interpretations/conclusions:
- The most likely cause of your missing data is a bad sensor, because over 70% of your raw values are 0

Having recently switched from Libre1 + bridge, to bridgeless Libre2, there is a lot of new information in the log, and I have found it rather like looking for a needle in a haystack, and have ended up asking dumb questions to the forum.

I am not aware of any section of the documentation dedicated to "How to understand the Event Logs" - but maybe creating this would be a useful precursor to any automation of the analysis. 
Apologies if this has been discussed before.

**Comment by Navid200** [comment]
There is a reasonable expectation that if a developer modifies a part of the code, they would accordingly edit the including logs that re affected.
This makes the maintenance of the logs a reasonable task.

But, if we create the capability for the logs to be interpreted, we will need an additional program or function.  Such a function or program will need to be maintained.  What happens when a developer modifies a feature in xDrip?  Whose responsibility will it be to update the interpreter?
I can already see a scenario where the interpreter becomes outdated and provides incorrect interpretations.  

Have you thought about that?
Or, am I misunderstanding your suggestion?

**Reply by nigelkeirross** [reply]
Thanks for your comment. I do not believe that you have misunderstood my suggestion. The overall idea is to increase the usefulness of the log messages, and broaden the audience who can benefit from these. 
As with any complex system, the interfaces should be well documented. I think this should also apply to the log messages. (I am not aware of a section of the user documentation that is dedicated to interpreting the log messages. If there is one, please can you send me the link). 
So any developer who adds new messages, or changes current ones should also document the format, and meaning & severity of the comment so that both users, and any developer of the automatic function can understand.

I certainly find messages in the log that I do not know if they indicate a problem, and if so, how serious, or what are the possible causes. For example a few hours ago I had a "LibreOOPAlgorithm" entry that started waitforUnlockPayload(sendGetBluetoothEnablePayload) got data payload is ... then about 9 screens of data. What does this indicate? 
Also there seems to be no consistent indication of the severity - is it normal behaviour / warning / unusual condition / error. Some errors are highlighted in red, but not all. 

As xDip+ spreads beyond the original very computer literate community to a broader user base, I think the need for easier troubleshooting will increase.

**Reply by Navid200** [reply]
Our (at least, my) thinking is that we provide documentation and support for all users.
Those are here:
https://xdrip.readthedocs.io/en/latest/
https://navid200.github.io/xDrip/

Someone who has difficulty using xDrip, should look at those.
The guides tell you how to use xDrip and how to troubleshoot.  
If there is a need to look at the logs, for troubleshooting, the guide tells you.

I support Dexcom.  If someone has difficulty using Dexcom with xDrip, they should not look at the logs in order to figure out what's going on as the first step.  They should look at the guide instead.

I personally do not support Libre.  @psonnera does.   

Even if we create a log interpreter, its maintenance will be a task that we do not have bandwidth for.  We are a few volunteers.  We need to use our time efficiently so that we can keep going instead of burning out.

If the guides/documentation linked above have any shortcomings, we need feedback.  Please tell us what is missing or what is not clear.

**Reply by nigelkeirross** [reply]
Thanks. The https://navid200.github.io/xDrip/ site was new to me. I agree with you that the documentation should be the starting point for troubleshooting. And certainly I'd agree that having comprehensive troubleshooting documentation is much more important than any tool I was proposing. I've looked over both links to see if I would have been able to diagnose the most recent problem I was having 3 sensors ago, which is what led me to looking at the logs, and asking questions on the forums. I haven't found anything yet, but will look more thoroughly over the next day or so.
From what I've seen, the troubleshooting & FAQs seem to address the "its not working at all" type of problems. 
Over the years most of my problems have been "why has is suddenly stopped working" or "why is it working so erratically" - like my last problem of working for say 90 minutes well, then off for 30 minutes, then good readings for a couple of hours, then 40 minutes of missed readings etc etc. Only by looking thru the logs & posting some, did I eventually diagnose the problem with expert help. 
I realise that libre is not your domain & @psonnera was very helpful. I have been a Libre1 user for 5 years or more, initially building my own version of the LimiTTer bridge, more recently using Bubble, and was fairly comfortable with dealing with missed reading issues etc. I am new the Libre2, and the ability to connect without a bridge, and this  so was very unfamiliar with how to troubleshoot the new system - and I couldn't find anything to help me.
It was this that led me to propose the "automatic log interpreter" idea. But I now think that it would be better to focus on getting better documentation on what the log messages mean, how to interpret them & how to diagnose problems from them.
I realise that some of this may well exist & I haven't found it yet. But I am willing to volunteer to help pull some of this together. Maybe a starting point would be for me to contact @psonnera and discuss if there is anything I could do to help.

**Reply by psonnera** [reply]
@nigelkeirross you'd be very welcome! https://xdrip.readthedocs.io/en/latest/troubleshoot/libre2/

**Reply by nigelkeirross** [reply]
@psonnera Thanks. I have looked at that link, and the ones that @Navid200 sent. Lots of good stuff. I will try to collect some thoughts about what I am looking for (and will be willing to try & create). I think it probably best that I contact you directly - what would be the best way to do that? Email or github? Thanks

**Reply by psonnera** [reply]
@nigelkeirross PM'd you

**Reply by nigelkeirross** [reply]
@psonnera I've looked on Github, gitlab & discord, but can't see any messages. Where should I look?

**Reply by psonnera** [reply]
You're right. Gitter moved to Matrix... Discord is a good idea.
